pennylane/kernels/cost_functions.py

```diff
@@ -0,0 +1,135 @@
+# Copyright
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains functionalities for kernel related costs.
+See `here <https://www.doi.org/10.1007/s10462-012-9369-4>`_ for a review.
+"""
+from pennylane import numpy as np
+from ..utils import frobenius_inner_product
+
+
+def square_kernel_matrix(X, kernel, assume_normalized_kernel=False):
+    """Computes the square matrix of pairwise kernel values for a given dataset.
+
+    Args:
+        X (list[datapoint]): List of datapoints
+        kernel ((datapoint, datapoint) -> float): Kernel function that maps datapoints to kernel value.
+        assume_normalized_kernel (bool, optional): Assume that the kernel is normalized, i.e.
+            that when both arguments are the same datapoint the kernel evaluates to 1. Defaults to False.
+
+    Returns:
+        array[float]: The square matrix of kernel values.
+    """
+    N = len(X)
+    matrix = [0] * N ** 2
+
+    for i in range(N):
+        for j in range(i, N):
+            if assume_normalized_kernel and i == j:
+                matrix[N * i + j] = 1.0
+            else:
+                matrix[N * i + j] = kernel(X[i], X[j])
+                matrix[N * j + i] = matrix[N * i + j]
```

Comment:
Just being curious, is there any advantage using a list over just creating `zeros` and filling it up?

Reply:
Yes, if you do it like this you can differentiate the whole result in a matrix sense. So for example if you have
```python
cost = np.linalg.inv(K)
```
where `K` is the kernel matrix you can differentiate it only if you build it like I do here with the list. At least that was the case last year when I did my project with the fisher info.

It has something to do with the way autograd handles those "boxes", if you do it like this you get a box that represents the matrix instead of an array of individual boxes.